pilot/pkg/model/push_context.go

```diff
@@ -78,9 +79,65 @@ type PushContext struct {
 	// Env has a pointer to the shared environment used to create the snapshot.
 	Env *Environment `json:"-,omitempty"`
 
+	// ServicePorts is used to keep track of service name and port mapping.
+	// This is needed because ADS names use port numbers, while endpoints use
+	// port names. The key is the service name. If a service or port are not found,
+	// the endpoint needs to be re-evaluated later (eventual consistency)
+	ServiceName2Ports map[string]map[string]uint32 `json:"-"`
+	ServicePort2Name  map[string]map[uint32]string `json:"-"`
+
 	initDone bool
 }
 
+// EDSUpdater is used for direct updates of the EDS model
+// and push.
+type EDSUpdater interface {
+	// EDSUpdateServieEntry is called when the list of endpoints or labels in a ServiceEntry is
+	// changed. For each cluster and hostname, the full list of active endpoints (including empty list)
+	// must be sent.
+	EDSUpdate(shard, hostname string, entry []*IstioEndpoint) error
```

Comment:
It is an action, it updates the entries used to compute EDS. Not really updating Endpoints - the endpoints are updated by k8s.

In general I want to avoid using "Endpoints" in methods dealing with EDS specific representation - so we don't confuse them with K8S Endpoints or service entry endpoints.